Sources/xcodeproj/Objects/Files/PBXFileElement.swift

```diff
@@ -141,15 +145,14 @@ public extension PBXFileElement {
     /// - Returns: file element absolute path.
     /// - Throws: an error if the absolute path cannot be obtained.
     public func fullPath(sourceRoot: Path) throws -> Path? {
-        let projectObjects = try objects()
         switch sourceTree {
         case .absolute?:
             return path.flatMap { Path($0) }
         case .sourceRoot?:
             return path.flatMap { sourceRoot + $0 }
         case .group?:
-            guard let group = projectObjects.groups.first(where: { $0.value.childrenReferences.contains(reference) }) else { return sourceRoot }
-            guard let groupPath = try group.value.fullPath(sourceRoot: sourceRoot) else { return nil }
+            guard let group = parent else { return sourceRoot }
```

Comment:
I have a concern with this. We are disregarding the fact that the user can add a child to a parent without calling the convenience methods where the call to `assignParentForChildren`. If that happens, this method would return an invalid result.

I'd rather fallback to the logic the we had before in case the file element doesn't have a parent.

Reply:
i'm added fallback. But i think it is not necessary because in all public methods for adding child called this method.

Reply:
Because of the fallback, the profit decreases, because the root group is searched (it has no parent and it falls into the fallback) - which is not fast if there are a lot of groups.

For my test project, this is 1 minute 30 seconds versus 3 seconds

What do you think about that ?

Reply:
> i'm added fallback. But i think it is not necessary because in all public methods for adding child called this method.

But it doesn't prevent us from adding a new method in the future and forgetting about this. It's good to add some safety here looking forward.

> Because of the fallback, the profit decreases, because the root group is searched (it has no parent and it falls into the fallback) - which is not fast if there are a lot of groups.

You could check whether the group is a root group and in that case return the `sourceRoot` path. What do you think?

Reply:
A good idea. But I do not understand how I can check it out?

Reply:
Some pseudocode of how I think you could achieve it:

```swift
let objects = try objects()
objects.projects.first({ $0.mainGroup == group })
```
It should be fast because objects typically contain one project. Let me know if that works :)